// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";
package observer;

import "structure_graph.proto";

service ObserverService {
  rpc CreateFuzzer(CreateFuzzerRequest) returns (CreateFuzzerResponse);

  rpc UpdateFeatures(UpdateFeaturesRequest) returns (UpdateFeaturesResponse);
}

message CreateFuzzerRequest {
  uint64 fuzzer_id = 1;
  StructureGraph structure_graph = 2;
}

message CreateFuzzerResponse {}

message UpdateFeaturesRequest {
  uint64 fuzzer_id = 1;

  message BitCounter {
    uint64 node_index = 1;
    uint32 counter = 2;
  }
  repeated BitCounter bit_counters = 2;

  uint64 corpus_id = 3;
}

message UpdateFeaturesResponse {
  message CorpusPriority {
    uint64 id = 1;
    uint32 priority = 2;
  }
  repeated CorpusPriority corpus_priorities = 1;
}
